x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java

```diff
@@ -239,6 +247,32 @@ public ValueFetcher valueFetcher(SearchExecutionContext context, String format)
         public IndexFieldData.Builder fielddataBuilder(FieldDataContext fieldDataContext) {
             throw new IllegalArgumentException("[semantic_text] fields do not support sorting, scripting or aggregating");
         }
+
+        public QueryBuilder semanticQuery(InferenceResults inferenceResults, float boost, String queryName) throws IOException {
+            String nestedFieldPath = name() + "." + CHUNKS;
+            String inferenceResultsFieldName = nestedFieldPath + "." + INFERENCE_CHUNKS_RESULTS;
+            QueryBuilder childQueryBuilder;
+
+            if (inferenceResults instanceof TextExpansionResults textExpansionResults) {
+                // TODO: Use WeightedTokensQueryBuilder
```

Comment:
We should use `WeightedTokensQueryBuilder` here directly. I haven't done that yet because we need to move that query builder (and its dependencies) to the `core` plugin first.

Reply:
PR for moving `WeightedTokensQueryBuilder`: https://github.com/elastic/elasticsearch/pull/106924

Reply:
It's not blocked anymore, I was planning on moving that one forward after finalizing this PR. I've already tested locally that using `WeightedTokensQueryBuilder` is functionally the same, so we can drop it in here in a follow-up PR.